RecoVertex/PrimaryVertexProducer/src/DAClusterizerInZT_vect.cc

```diff
@@ -174,143 +360,188 @@ double DAClusterizerInZT_vect::update(
+    gvertices.swt_ptr[ivertex] = 0.0;
+    gvertices.szz_ptr[ivertex] = 0.0;
+    gvertices.stt_ptr[ivertex] = 0.0;
+    gvertices.szt_ptr[ivertex] = 0.0;
+#ifdef USEVTXDT2
+    gvertices.sumw_ptr[k] = 0.0;
+#endif
   }
 
   // loop over tracks
   for (auto itrack = 0U; itrack < nt; ++itrack) {
-    kernel_calc_exp_arg(itrack, gtracks, gvertices);
-    local_exp_list(gvertices.ei_cache_, gvertices.ei_, nv);
+    unsigned int kmin = gtracks.kmin[itrack];
+    unsigned int kmax = gtracks.kmax[itrack];
+
+#ifdef DEBUG
+    assert((kmin < kmax) && (kmax <= nv));
+    assert(itrack < gtracks.Z_sum.size());
+#endif
 
-    gtracks.Z_sum_[itrack] = kernel_add_Z(gvertices);
-    if (edm::isNotFinite(gtracks.Z_sum_[itrack]))
-      gtracks.Z_sum_[itrack] = 0.0;
+    kernel_calc_exp_arg_range(itrack, gtracks, gvertices, kmin, kmax);
+    local_exp_list_range(gvertices.ei_cache_ptr, gvertices.ei_ptr, kmin, kmax);
+
+    gtracks.Z_sum_ptr[itrack] = kernel_add_Z_range(gvertices, kmin, kmax);
+    if (edm::isNotFinite(gtracks.Z_sum_ptr[itrack]))
+      gtracks.Z_sum_ptr[itrack] = 0.0;
     // used in the next major loop to follow
-    sumpi += gtracks.pi_[itrack];
+    sumpi += gtracks.pi_ptr[itrack];
 
-    if (gtracks.Z_sum_[itrack] > 1.e-100) {
-      kernel_calc_normalization(itrack, gtracks, gvertices);
+    if (gtracks.Z_sum_ptr[itrack] > 1.e-100) {
+      kernel_calc_normalization_range(itrack, gtracks, gvertices, kmin, kmax);
     }
   }
 
   // now update z, t, and pk
-  auto kernel_calc_zt = [sumpi,
-                         nv
-#ifdef VI_DEBUG
-                         ,
-                         this
-#endif
-  ](vertex_t& vertices) -> double {
+  auto kernel_calc_zt = [sumpi, nv](vertex_t& vertices) -> double {
     double delta = 0;
 
     // does not vectorizes
     for (unsigned int ivertex = 0; ivertex < nv; ++ivertex) {
-      if (vertices.nuz_[ivertex] > 0.) {
-        auto znew = vertices.swz_[ivertex] / vertices.nuz_[ivertex];
-        // prevents from vectorizing if
-        delta += std::pow(vertices.z_[ivertex] - znew, 2);
-        vertices.z_[ivertex] = znew;
+      if (vertices.nuz_ptr[ivertex] > 0.) {
+        auto znew = vertices.swz_ptr[ivertex] / vertices.nuz_ptr[ivertex];
+        delta = max(std::abs(vertices.z_ptr[ivertex] - znew), delta);
+        vertices.z_ptr[ivertex] = znew;
       }
 
-      if (vertices.nut_[ivertex] > 0.) {
-        auto tnew = vertices.swt_[ivertex] / vertices.nut_[ivertex];
-        // prevents from vectorizing if
-        delta += std::pow(vertices.t_[ivertex] - tnew, 2);
-        vertices.t_[ivertex] = tnew;
+      if (vertices.nut_ptr[ivertex] > 0.) {
+        auto tnew = vertices.swt_ptr[ivertex] / vertices.nut_ptr[ivertex];
+        //delta = max(std::abs(vertices.t_ptr[ ivertex ] - tnew), delta); // FIXME
+        vertices.t_ptr[ivertex] = tnew;
+#ifdef USEVTXDT2
+        vertices.dt2_ptr[ivertex] = vertices.nut_ptr[ivertex] / vertices.sumw_ptr[ivertex];
+#endif
+      } else {
+        // FIXME
```

Comment:
Which is the status of this expected fix?

Reply:
this and other pieces of code with #ifdef USEVTXDT2 are supposed to eventually address the situation that arises when a cluster consists entirely of tracks without timing information.  In that case it get's a bogus time value which may prevent associating tracks with timing information at a later stage.
I didn't have the time to test this carefully so I left it out for now. It is not addressed in the original code either. Technically it does not happen there, because all tracks enter every cluster, albeit with tiny weights. So effectively it has always been like that an the new code does not represent a change in this respect.